### ISSUE TITLE
Stunt Race FX Models

### ISSUE THREAD

**Issue by mills32** [Hordeking/sfxObjectDump#1]
Hi!

I just discovered this, and I wanted to ask if somebody knows how to rip the 3d models from Stunt Race FX.

Thanks!

**Comment by Hordeking** [comment]
I haven't done much but yank the textures from Stunt Race (they're horz mirrored for some reason and use a different palette). I imagine the model format for them would be similar to that in starfox/starfox 2.

If you can find the object list that has the offsets to the vertex list, poly list, scaling factor and palette used, I will document those and put them up, though.

My code has a "validation" function, but for now, it really doesn't do anything useful. Optimally, I would like to be able to use that to scan a file starting at some offset and determine if something is a valid object or not. I think that would be best done by having one for the vertex list (failing if it finds something that shouldn't be in there, essentially parsing the list), and one for the poly list.

Congratulations. I think you're the first to find this code. I've been tinkering with this since '13, and finally got ahold of enough information to really parse the data about a year or two ago.

PS: These games sometimes build objects and enemies from several sub-objects, for instance plasma hydra, Monarch Dodora, and the rock crusher spring to mind. I don't know how the game does this, but I suspect it's probably done using the level definition data. If you want to help me hack at this stuff, I'd love the extra hand at it. For the moment, I'm trying to completely parse out the palette data into material files so I can output fully texture objects (one file per keyframe, yes, I have produced untextured fully animated objects in blender from these).

PS2: Vertex lists and poly lists are separate. Any pair of vertex list and poly list with the same number of vertices will actually work together without breaking anything. But if the poly list doesn't go to the vertex list, boy do they look weird. This was probably done so they could have multiple iterations of various objects, like shadow-arwings, wireframe ones, etc with different palettes, but the same BSP tree and poly list.

**Comment by mills32** [comment]
I'd love to help, but I'm not very experienced in hacking. 
How can I start to look for that offsets (vertex, poly, scaling, palette)?.
I read some info about starfox but I can't find that web page again.

I'm interested in ripping the tracks and cars from stunt race because I wanted to make a little "port" (just a car and a track) and a video showing it running at 60 fps :).

**Comment by Hordeking** [comment]
The tracks themselves might be height maps (just a wild guess), so if you went looking for images/textures in the rom, you might be able to spot it. I found something like that in the Dirt Trax rom (which I can't get past the menu screens).

I have gone out of my way to save a lot of the information I found since '13. I also did some of my own original documentation. I don't know if there's a way to simply dump information documents on github, but I think maybe a google drive would work for this. My own docs, those would be easy to just stick on github. The webpage dumps, not quite as easy.

At the moment, I'm working on the palette/texture information stuff, since I have mostly gotten the models down to a known fact (I need to do the verify function so I can exhaustively determine if there are any other vertex types that I haven't documented).

If you're looking to do a port, you've got a long way to go. That's a lot of programming right there. If you're interested in learning how to use blender, we can probably do something from there.

**Comment by Hordeking** [comment]
To answer your question about finding the models and stuff, VL-Tone apparently used a debugger to figure it out. I don't know how to use a debugger well enough to find things like that, unfortunately. I tend to reverse engineer static data from the executables/roms.

If you want to learn, my understanding is that vSNES is the go-to debugger, and can analyze save states and everything. That means you should be able to find your models in the ram somewhere. If you find anything, please don't hesitate do document it (where it was found, rom md5 checksum so the same rom can be used, address of the beginning of the data, and possibly even a dump of the actual data itself so it can be found later if the address isn't fixed, etc.)

**Comment by mills32** [comment]
Thanks! 
I dont really want to make a port, I'll just use something like blender game engine to drive a car in ripped tracks for fun :).

**Comment by Hordeking** [comment]
That would probably be simplest. But if you're interested in helping, I would be glad to make a record of your findings and incorporate changes via pull requests. There's a reasonable chance that Stunt Race FX does have some differences.

**Comment by Hordeking** [comment]
In case you're up for it, here's a render of one of the models from the original star fox. In 3d side by side.
[Star Fox Model Test](https://www.youtube.com/watch?v=EPjDhvRAG-8)

**Comment by mills32** [comment]
It looks like somebody found where in the ROM is the model data: https://www.smwcentral.net/?p=viewthread&t=23423&page=1&pid=343034#p343034.

**Comment by Hordeking** [comment]
It looks like Matrizzle again. He appears to be pretty good at superfx hacking. Anyway, the data he's posting looks like the pointers to the object header data (like the list in star fox that points to the various object headers, which in turn do things like point to the vertex data, face data, scale, palette, etc). Feel free to dig down into this data and let me know where everything is. I'm teaching right now, so I don't have a lot of time at the moment, but once I have time, I might play with it.

He may have figured out what the correct palette was for the textures? I dunno, but I also found them at 0x30200 (rom with header, 0x30000 with no header).

**Comment by zoncabe** [comment]
hi i have a doubt about the second field to be added in the usage string

sfxobjdump 'romfile' 'vertex list address' 'face data address'

we know 'vertex list address' is posted by Matrizzle but..
how do i know 'face data address'??

**Comment by Hordeking** [comment]
@zoncabe I would have encouraged you to write this as a new post, but I will briefly answer it here. If you need further assistance, open a new thread so it's visible to others looking for it. In short, a complete model  in star fox is composed of three general data structures: the vertex list, the face data, and texture data. This code does not yet texture the palette stuff as I haven't been able to find a good way to export the animated stuff (maybe as a gif, I dunno). Skip to the end if this is all review for you.

The vertex list is exactly that, a list of vertices that the model uses. If you plotted these as a point cloud in Blender, you'd kind of be able to make out the shape of whatever it is. The face data is the other part of the picture, basically this is what tells us which points are taken together as shapes. This is our actual polygon data. It also handles things like what color or animation from the selected palette is used for that shape.

A simple example: a vertex list with three points, and a face list with one entry that uses those three points. The game engine would draw not three individual points, but a triangle. Verticess can be used in multiple faces at once.

Star Fox objects _usually_ have a vertex list followed immediately by the face data. However, let's say we want to draw a regular arwing and a wireframe one (like in the training mode). Rather than repeating the vertex list, since the same list is used for both models, it has a lump of vertex data in one place and reuses it for both models. When it wants to draw the solid one it grabs the face data (with associated color entries and so on) and renders that. When it wants the wireframe one it uses the face data for the wire one. Both objects have the same number of vertices, and the face/edge data will still use that list.

Long story short, there's a list of object structures early on in the rom that contains pointers to the vertex data, the face data, and some other stuff for scaling. Usually the face data comes immediately after the vertex data, but not always. So you need to know both addresses.

I've uploaded [my document](https://drive.google.com/file/d/1KeQXLywM3CBam02HJpXJlJKquYoPb8oI/view?usp=sharing) that details all of this stuff. It seems to be mostly correct. I derived all of the information from the object structure data stored in the rom. One thing you should always be aware of is whether the rom you have is headered or not. If it is, when putting in the address, you would need to account for the header when giving it an offset.
